daemon/cmd/ipcache.go

```diff
@@ -102,91 +101,163 @@ func containsSubnet(outer, inner net.IPNet) bool {
-// (Aside: prefix references mostly come from network policies, either directly through CIDR
-// selectors or via ToFQDN rules.)
+// The purpose of this is to preserve stable local identities on agent
+// restart as much as possible. This helps prevent spurious policy drops
+// on agent restart.
 //
-// For ingress IPs, it will add those to the ipcache and configure the local node
-// accordingly.
+// After a grace period, the restored identity references and placeholder ipcache
+// metadata entries are removed, assuming that the agent has synchronized
+// with other state (i.e. kvstore, k8s) and that all necessary entries
+// are present in ipcache & the identity allocator.
 //
 // This *must* be called before initMaps(), which will hide the "old" ipcache.
-func (d *Daemon) restoreIPCache() error {
-	ingressIPs := make([]netip.Prefix, 0, 2)
-	// need to preserve this so we can remove it later.
-	d.restoredCIDRs = map[netip.Prefix]identity.NumericIdentity{}
+func (d *Daemon) restoreLocalIdentities() error {
+	// Restore the local identity allocator from its checkpoint.
+	// This returns the set of identities created. We will use this set
+	// to regenerate the set of labels for prefixes in the ipcache.
+	restoredIdentities, err := d.identityAllocator.RestoreLocalIdentities()
+
+	// Dump the existing BPF ipcache map
+	localPrefixes, err2 := d.dumpOldIPCache()
+	if err2 != nil {
+		log.WithError(err2).Warn("Failed to restore existing identities from the previous ipcache. This may cause policy interruptions during restart.")
+		err = errors.Join(err, err2)
+		// continue; we may have a partial dump
+	}
+
+	// create placeholder CIDR labels in the ipcache.
+	// This only adds an ipcache metadata entry for prefixes with a `cidr:`
+	// label and ingress IPs. All other entries will have to be created anew.
+	d.restoreIPCache(localPrefixes, restoredIdentities)
+	return err
+}
+
+// dumpOldIPache reads the soon-to-be-overwritten ipcache BPF map, noting any prefixes
+// with a locally-scoped or ingress identity.
+func (d *Daemon) dumpOldIPCache() (localPrefixes map[netip.Prefix]identity.NumericIdentity, err error) {
+	localPrefixes = map[netip.Prefix]identity.NumericIdentity{}
 
 	// Dump the bpf ipcache, recording any prefixes with local or ingress
 	// numeric identities.
-	err := ipcachemap.IPCacheMap().DumpWithCallback(func(key bpf.MapKey, value bpf.MapValue) {
+	err = ipcachemap.IPCacheMap().DumpWithCallback(func(key bpf.MapKey, value bpf.MapValue) {
 		k := key.(*ipcachemap.Key)
 		v := value.(*ipcachemap.RemoteEndpointInfo)
 		nid := identity.NumericIdentity(v.SecurityIdentity)
 
-		if isLocalIdentity(nid) {
-			d.restoredCIDRs[k.Prefix()] = nid
-		} else if nid == identity.ReservedIdentityIngress && v.TunnelEndpoint.IsZero() {
-			ingressIPs = append(ingressIPs, k.Prefix())
+		if nid.Scope() == identity.IdentityScopeLocal || (nid == identity.ReservedIdentityIngress && v.TunnelEndpoint.IsZero()) {
```

Comment:
Why do we no longer use `isLocalIdentity` here? Was the previous check for `IdentityScopeRemoteNode` to preserve the numeric identity, but not the IPCache entry?

Reply:
Good question!

There's no longer any point in restoring node entries directly in to the ipcache, since we'll just pick up the same numeric ID when re-upserting the nodes in from k8s / kvstore. In fact, doing so is kind of a bad idea, as IPs could "stick" as being node IPs for up to 10 minutes incorrectly.